### PR TITLE
[RFC] [Testcase Refactoring]Make distributed replicate training tests device-agnostic

### PR DESCRIPTION
**Pull request by zhaiyaxing001**
## Summary

This PR refactors `test/distributed/_composable/test_replicate_training.py` to be device-agnostic by removing the global `device_type` variable, splitting HPU-incompatible tests into a separate class, and adding `instantiate_device_type_tests` with `HardwareClassification` tags.

## Motivation

The original test had hardcoded device assumptions:
- Module-level global `device_type = torch.device(get_devtype())`
- `@unittest.skipIf(TEST_HPU, ...)` decorators for sleep kernel tests
- Device type validation `if test_device_type not in ("cuda", "hpu", "xpu", "cpu")`
- No `device` parameter, no `hw_classification`, no `instantiate_device_type_tests`

### 1. Remove global `device_type` variable and imports

```diff
-from torch.testing._internal.common_fsdp import get_devtype
-device_type = torch.device(get_devtype())
-import unittest
-    TEST_HPU,
```

### 2. Add imports

```diff
+from torch.testing._internal.common_device_type import instantiate_device_type_tests
 from torch.testing._internal.common_utils import (
+    HardwareClassification,
     ...
 )
```

### 3. Add `hw_classification` (10 classes)

All classes set `hw_classification = HardwareClassification.ACCELERATOR`.

### 4. Class split

**`TestReplicateCastAfterInit` → renamed to `TestReplicateCastAfterInitPrecision`**

**`TestReplicate1DTrainingCore` → split into two classes:**

| Retained in `TestReplicate1DTrainingCore` | Moved to `TestReplicate1DTrainingCoreNoHpu` |
|-------------------------------------------|--------------------------------------------|
| `test_train_parity_single_group` | `test_train_parity_multi_groups` |
| `test_non_root_forward_backward` | `test_train_parity_multi_group_cpu_offload_eager` |
| `test_multi_forward_module` | `test_post_optim_event` |
| `test_explicit_prefetching` | |

`TestReplicate1DTrainingCoreNoHpu` uses `except_for=["cpu", "hpu"]` (contains sleep kernel tests).

### 5. Add `device` parameter (16 methods)

All test methods get `device` parameter.

### 6. Add `instantiate_device_type_tests` (10 classes)

```python
instantiate_device_type_tests(TestReplicateForwardInputs, globals(), except_for="cpu", allow_xpu=True)
instantiate_device_type_tests(TestReplicateRegisteredParams, globals(), except_for="cpu", allow_xpu=True)
instantiate_device_type_tests(TestReplicateCastAfterInitPrecision, globals(), except_for="cpu", allow_xpu=True)
instantiate_device_type_tests(TestReplicate1DTrainingCore, globals(), except_for="cpu", allow_xpu=True)
instantiate_device_type_tests(TestReplicate1DTrainingCoreNoHpu, globals(), except_for=["cpu", "hpu"], allow_xpu=True)
instantiate_device_type_tests(TestReplicateTrainingCompose, globals(), except_for="cpu", allow_xpu=True)
instantiate_device_type_tests(TestReplicateSharedParams, globals(), except_for="cpu", allow_xpu=True)
instantiate_device_type_tests(TestReplicateGradientAccumulation, globals(), except_for="cpu", allow_xpu=True)
instantiate_device_type_tests(TestReplicateCustomForwardMethod, globals(), except_for="cpu", allow_xpu=True)
instantiate_device_type_tests(TestReplicateTPTraining, globals(), except_for="cpu", allow_xpu=True)
```

## Testing

- **NPU**: 6 tests skipped (4 sleep + 2 fp64), 11 tests passed
- **HPU**: Behavior unchanged
- **CUDA**: Behavior unchanged
- **CPU**: All tests skipped (expected)

## Compatibility

- Uses `torch.accelerator.current_accelerator()` public API
- Maintains CUDA/HPU behavior unchanged
- Note: NPU has a specific DTensor.to() limitation that requires a device-type check

## Notes

1. This PR depends on the following PR code changes(in OPEN state):
image

2. The NPU does not support the `_sleep` kernel. An issue has been submitted. After the issue is resolved, the corresponding skip can be enabled. The issue link is as follows: https://gitcode.com/Ascend/pytorch/issues/3958.

### CODE REVIEW

File: test/distributed/_composable/test_replicate_training.py [hunk 1]
```diff
@@ -198,14 +212,19 @@ def _assert_same_params(
...
     def world_size(self) -> int:
         return 2
 
     @skip_if_lt_x_gpu(1)
     @wrapSwapTensorsTest(True)
+    @requires_capabilities(Capability.distributed.fsdp)
     def test_to_float64_after_init(self):
         """Tests that the user can cast the module to float64 after init."""
+        if device_type.type == "npu":
+            self.skipTest("NPU backend does not fully support DTensor.to() for dtype conversion")
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py]: If this test case is not available on npu. How about use blacklist mechanism(yaml or json file) to filter this case.

File: test/distributed/_composable/test_replicate_training.py [hunk 2]
```diff
@@ -454,6 +478,8 @@ def delayed_reduce_scatter(*args, **kwargs):
                 self.assertEqual(losses[0], losses[1])
 
     @skip_if_lt_x_gpu(2)
+    @unittest.skipIf(_should_skip_for_accelerator(), "Sleep kernel not supported for NPU/HPU")
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py]: NPU is out-of-tree device. So we could use blacklist mechanism to filter this testcase.

File: test/distributed/_composable/test_replicate_training.py (hunk 1 above)
Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py]: It has been changed to ACCELERATOR.
instantiate_device_type_tests excep_for npu

File: test/distributed/_composable/test_replicate_training.py (hunk 2 above)
Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py]: The related methods have been extracted as a separate class and modified to ACCELERATOR.
Instantiate the device type tests to exclude the NPU.

File: test/distributed/_composable/test_replicate_training.py [hunk 3]
```diff
@@ -253,7 +271,12 @@ def test_to_float64_after_init(self):
                 _optim.zero_grad(set_to_none=(iter_idx % 2 == 0))
 
 
+instantiate_device_type_tests(TestReplicateCastAfterInitPrecision, globals(), except_for=["cpu", "npu"])
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py]: `npu` should be removed here. We could use blacklist to filter this test case on npu device.

File: test/distributed/_composable/test_replicate_training.py [hunk 4]
```diff
@@ -367,7 +497,7 @@ def _test_train_parity_multi_group(
             in (2, 3)
         ):
             return
-        if test_device_type not in ("cuda", "hpu", "xpu", "cpu"):
+        if test_device_type not in ("cuda", "hpu", "xpu", "npu", "cpu"):
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py]: Adding `npu` here is not appropriate. This if statement means that current test case only works on accelerator, we can use `onlyAccelerator` annotation or `instantiate_device_type_tests` to realize.

File: test/distributed/_composable/test_replicate_training.py [hunk 5]
```diff
@@ -640,14 +670,17 @@ def step_post_hook(
...
             losses[-1].backward()
             optim.step()
-            # Sleep after the optimizer step to allow CPU to run ahead into the
-            # next iteration's forward, exercising the post-optim stream sync
             torch.get_device_module(device_type)._sleep(int(25 * get_cycles_per_ms()))
         for ref_loss, loss in zip(ref_losses, losses):
             self.assertEqual(ref_loss, loss)
 
 
+instantiate_device_type_tests(TestReplicate1DTrainingCoreSleep, globals(), except_for=["cpu", "hpu", "npu"])
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py]: It's better to remove `npu`  here.

File: test/distributed/_composable/test_replicate_training.py (hunk 3 above)
Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py]: It has been removed. The blacklist mechanism will be used for subsequent processing.

File: test/distributed/_composable/test_replicate_training.py (hunk 5 above)
Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py]: It has been removed. The blacklist mechanism will be used for subsequent processing.

File: test/distributed/_composable/test_replicate_training.py (hunk 4 above)
Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py]: The values are duplicate and have been removed. used the instantiation device type test to control them.

File: test/distributed/_composable/test_replicate_training.py [hunk 6]
```diff
@@ -96,15 +92,20 @@ def forward(self, x: torch.Tensor, ys: tuple[torch.Tensor, ...]):
...
     @property
     def world_size(self) -> int:
         return 4
 
     @skip_if_lt_x_gpu(1)
-    def test_param_registration_after_forward(self):
+    def test_param_registration_after_forward(self, device):
         """Tests the parameter registration after forward."""
-        device = torch.device(device_type.type, 0)
+        device = torch.device(self.device_type)
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py:108]: plz check whether we should use input parameter `device` or `self.device_type`? Is `self.device_type` injected by `instantiate_device_type_tests`?

File: test/distributed/_composable/test_replicate_training.py [hunk 7]
```diff
@@ -358,7 +475,6 @@ def _test_train_parity_multi_group(
         delay_before_optim: bool,
         unshard_async_op: bool,
     ):
-        # Only test individual delays or all four delays to save test time
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py:361]: Is there no need to delete this comment?

File: test/distributed/_composable/test_replicate_training.py (hunk 6 above)
Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py:108]: The injected device is the same as the value of self.device_type, which is set in setUpClass().

File: test/distributed/_composable/test_replicate_training.py (hunk 7 above)
Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py:361]: The comment was mistakenly deleted and has been re-added.

File: test/distributed/_composable/test_replicate_training.py [hunk 8]
```diff
@@ -454,22 +569,22 @@ def delayed_reduce_scatter(*args, **kwargs):
                 self.assertEqual(losses[0], losses[1])
 
     @skip_if_lt_x_gpu(2)
-    def test_non_root_forward_backward(self):
+    def test_non_root_forward_backward(self, device):
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py]: In original logic, there is no `skip TEST_HPU` annotation on this method. So is it right to put `test_non_root_forward_backward` in class TestReplicate1DTrainingCoreSleep?

File: test/distributed/_composable/test_replicate_training.py [hunk 9]
```diff
@@ -304,17 +315,124 @@ def _test_train_parity_single_group(self, lin_shapes: list[tuple[int, int]]):
...
+                _optim.zero_grad()
+                _optim.step()
+
+            self.assertEqual(losses[0], losses[1])
+
+
+instantiate_device_type_tests(TestReplicate1DTrainingCore, globals(), except_for="cpu", allow_xpu=True)
+
+
+class TestReplicate1DTrainingCoreSleep(FSDPTest):
```
Comment (wjlFlyer) [inline test/distributed/_composable/test_replicate_training.py]: class name `TestReplicate1DTrainingCoreSleep` is a bit confused. Do you mean that put the methods with annotation `skip TEST_HPU` in this class? How about name it `TestReplicate1DTrainingCoreNoHpu`?

Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py]: The method name has been modified.

File: test/distributed/_composable/test_replicate_training.py (hunk 8 above)
Comment (zhaiyaxing001) [reply test/distributed/_composable/test_replicate_training.py]: This method does not have an HPU skip label, does not require moving to a new class, and has been processed.

**Review by wjlFlyer — APPROVED** [review]

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill
